core/CronArchive.php

```diff
@@ -561,663 +418,312 @@ public function run()
         $this->logger->info($timer->__toString());
     }
 
-    public function getErrors()
+    private function isDoneFlagForPlugin($doneFlag)
     {
-    	return $this->errors;
+        return strpos($doneFlag, '.') !== false;
     }
 
-    /**
-     * End of the script
-     */
-    public function end()
+    private function archiveArrayContainsArchive($archiveArray, $archive)
     {
-        /**
-         * This event is triggered after archiving.
-         *
-         * @param CronArchive $this
-         */
-        Piwik::postEvent('CronArchive.end', array($this));
-
-        if (empty($this->errors)) {
-            // No error -> Logs the successful script execution until completion
-            Option::set(self::OPTION_ARCHIVING_FINISHED_TS, time());
-            return;
-        }
-
-        $this->logSection("SUMMARY OF ERRORS");
-        foreach ($this->errors as $error) {
-            // do not logError since errors are already in stderr
-            $this->logger->info("Error: " . $error);
+        foreach ($archiveArray as $entry) {
+            if ($entry['idsite'] == $archive['idsite']
+                && $entry['period'] == $archive['period']
+                && $entry['date1'] == $archive['date1']
+                && $entry['date2'] == $archive['date2']
+                && $entry['name'] == $archive['name']
+            ) {
+                return true;
+            }
         }
-
-        $summary = count($this->errors) . " total errors during this script execution, please investigate and try and fix these errors.";
-        $this->logFatalError($summary);
+        return false;
     }
 
-    public function logFatalError($m)
+    private function getNextInvalidatedArchive($periodToGet)
     {
-        $this->logError($m);
+        $tables = $this->getTablesWithInvalidatedArchives();
 
-        throw new Exception($m);
+        foreach ($tables as $table) {
+            $tableMonth = substr($table, strlen($table) - 7, 7);
+            $tableMonth = str_replace('_', '-', $tableMonth);
+
+            if (!isset($this->idArchivesToExclude[$tableMonth])) {
+                $this->idArchivesToExclude[$tableMonth] = [];
+            }
+
+            $nextArchive = $this->model->getNextInvalidatedArchive($table, $periodToGet, $this->allWebsites, $this->idArchivesToExclude[$tableMonth]);
```

Comment:
> Usually, the first archiver would work on site1, the next archiver would work on site 2, and the next archiver would work on site 3.

Not into the code, but wanted to agree that this feature where each concurrent archiver archives a different site is really valuable. Why? often one website has lots of segments and is very slow, while 10 or 50 other sites will be quick to archive. And if we didn't process them concurrently then you would end up potentially having to wait in some cases many hours to see data for "yesterday" for many/most of the sites.

Reply:
> Then there's some risk that these would be worked on first before it archives any other site?

Yes, this is a risk. We could go one site at a time instead of looking for any archive from any site. Eg, first we select archives only for iSite=1, then idSite=2, etc. Multiple archivers might start pulling from the same site, but if we move on from idSite=1 and a segment is created for it, archivers that have gone past it will ignore it until they execute again. Though we'll still have to go through those days if the invalidation occurs while archiving idSite=1. This might make the `done%` query a bit faster too...

We could also maybe save the start time of archiving and only select invalidated archives where `ts_archived < start time`? So if an archive is written after archiving started, and then invalidated, we don't pick it up again.

Reply:
> Why? often one website has lots of segments and is very slow, while 10 or 50 other sites will be quick to archive.

Maybe not in this PR, but does it sound like a good idea to save the elapsed time of an archive, then order the idSites by shorted elapsed time for yesterday's archive? This way we archive smaller sites first.

Reply:
> Maybe not in this PR, but does it sound like a good idea to save the elapsed time of an archive, then order the idSites by shorted elapsed time for yesterday's archive? This way we archive smaller sites first.

Ideally it would keep the current behaviour where each process grabs a different site ID and works through it, picking the next one then. That works quite well and simple. 

Another simple idea could be to pick first the websites that have less segments (ordered by idsite ASC since usually important sites were created first), as we expect them to archive sooner in most cases.

Reply:
> We could also maybe save the start time of archiving and only select invalidated archives where ts_archived < start time? So if an archive is written after archiving started, and then invalidated, we don't pick it up again.

That be a good start.

> Another simple idea could be to pick first the websites that have less segments (ordered by idsite ASC since usually important sites were created first), as we expect them to archive sooner in most cases.

That won't really be much helpful as the number of segments doesn't say much how long they take etc.

I guess in general we could get the list of idSites and iterate over them and always get a different site but that might make things quite complicated and could still cause random issues (also we'd need to iterate per idSite per archive table etc it could become complicated eg remember which archive table we looked at last time for this specific site and a different archiver might have in the meantime done something else etc). 

I suppose keeping the old behaviour of `SharedSiteIds` might work best overall indeed and could indeed also make some archive table queries faster.

Reply:
Just had an idea that might fix a bunch of issues, but haven't thought it through completely, so not 100% sure:

- we add a new column to archive tables, `invalidation_time` and add an index on it
- when looking for archives to re-archive we look for archives `ORDER BY invalidation_time ASC`, because of the index it ends up being a fast query

At this point we're pulling archives in order, so we sort of go site by site. However, if we want to have one process archiving one site at a time, we can do:

- loop through each idSite
- for each idSite, try to lock on `CronArchive.siteArchive.$idSite`, if it works, archive for site
- if not, try next idSite
- (after each climulti run expire lock for 5 hours, hopefully shouldn't add too many expire calls since it's one per N climulti:requests)
- after a site is done archiving, reset invalidated tables and re-query for next site (due to index added above, shouldn't be that bad?)

What do you think of each of these?

Reply:
> , try to lock on CronArchive.siteArchive.$idSite, if it works, archive for site

not sure what you mean? You mean create a lock for that? Ideally, if any possible, it be great to avoid locks at (almost all) cost since they often cause issues and are later hard to debug what is causing issues. I suppose you mean that is to check if another process is already working on this. A problem may be again that we need quite long expiry times since archiving can take many hours and we even have some single queries here that can take 3-4 hours or even longer. Meaning we can't extend that lock easily and if we set to lock to say 12 hours, and for some reason it doesn't get unlocked, then nothing happens for 12 hours for that site... 

Also generally not sure re `invalidation_time ASC` as it means when a new segment is created, and many entries with same invalidation time are created, then it will continue to archive on those segments for quite a while until it goes to the next. Considering we'd still iterate through the other sites it might work though. 

> loop through each idSite

As mentioned earlier, a tricky part is remember which archive table you looked at last for each site: 

`also we'd need to iterate per idSite per archive table etc it could become complicated eg remember which archive table we looked at last time for this specific site and a different archiver might have in the meantime done something else etc`

Trying to think what behaviour we'd generally want. I suppose generally, we want to archive most recent dates first meaning we prefer archiving today and yesterday for example over a date that is 3 or 5 days back. 

Generally, I suppose we also prefer all idSites to be archived regularly and older dates can be archived "later" once it actually gets to it. At the same time, we don't want to run into an issue where it never gets to archive these older dates because it's always busy archiving recent dates. So I suppose there has to be trade offs.

Would there be anything wrong that we'd add back this `SharedSiteIds` @diosmosis ? It seems this worked quite well and made sure we distribute things quite well and should there be one site with many segments to rearchive, then one worker would focus on that, while others focus on the next sites... It's of course not perfect either

Reply:
> Also generally not sure re invalidation_time ASC as it means when a new segment is created, and many entries with same invalidation time are created, then it will continue to archive on those segments for quite a while until it goes to the next. Considering we'd still iterate through the other sites it might work though.

Maybe we could ignore new segments at first, then do that check after the initial command is run, and if needed, do a last7 query or something? After the entire core:archive run.

Reply:
> Would there be anything wrong that we'd add back this SharedSiteIds @diosmosis ?

@tsteur I'm only really concerned w/ adding back the code, not the approach. The lock per site approach I described should have the same effect but require less code? I'm thinking the main issue w/ locks in the other issue was just that it was re-expired too often (after every query). There we re-expire every 30m or more now, and we could do the same thing w/ idSites? We'd have to keep track of the archive tables/site combo's but that can be taken care of w/ an associative array, no?

Reply:
> The lock per site approach I described should have the same effect but require less code?

Not sure. Personally, I'd just like to avoid any locks since pretty much all problems we have now are related to locks (in code or DB internal locks). 

It could work to iterate over every site but would maybe not use locks. Also I understand you would iterate over each site within `core:archive`? I don't think a lock would be really needed and it be fine if multiple concurrently running archivers work on the same site should they move on to the next site afterwards.

> We'd have to keep track of the archive tables/site combo's but that can be taken care of w/ an associative array, no?

Yeah should be possible. Was just thinking it might create some additional queries etc eg 
* Archiver 1, idSite = 1, works on archive_numeric_2020_02
* Archiver 2, idSite = 1, works on archive_numeric_2020_03
* Then Archiver 1, idSite = 1, would next work on archive_numeric_2020_03 instead of directly going to `archive_numeric_2020_04` maybe? 

Might not be much of an issue though as it would mean only additional queries on archive tables that might not return a result. Not sure if it can cause any other issues if we don't use locks.

Reply:
> I suppose keeping the old behaviour of SharedSiteIds might work best overall indeed and could indeed also make some archive table queries faster.

Was there an issue with the `SharedSiteIds`? this part of the archiver afaik never had any issue and worked good and did the job well.

Reply:
@mattab it's just more code than I think is needed to get the same result.